Provide better services for incomplete generic calls

andy-ms opened the pull request:
Fixes #15412

* Add a `getBestGuessSignature` method to the checker which acts like `getResolvedSignature`, but which does a better job in the case of an error.

* Normally the type checker will automatically fill in `{}` for type arguments that it can't infer. We now track when we do this, so if we have a Signature we can tell whether it has a bogus instantiation. Would also be useful for [`no-inferred-empty-object-type`](https://palantir.github.io/tslint/rules/no-inferred-empty-object-type/) which currently has an unreliable implementation.

mhegazy commented:
@andy-ms can you check with @ahejlsberg, he had thoughts about this area for a similar change.

andy-ms commented on src/compiler/checker.ts:
```diff
@@ -69,6 +69,9 @@ namespace ts {
         const argumentsSymbol = createSymbol(SymbolFlags.Property, "arguments");
 
+        /** This will be set during calls to `getResolvedSignature` where services determines an apparent number of arguments greater than what is actually provided. */
+        let apparentArgumentCount: number | undefined;
```
It would be nice if we just computed this at the moment we need it, but that seems to involve getting the tokens array, which is currently code exclusive to services. See `getArgumentCount` in `signatureHelp.ts`.

andy-ms commented:
Per discussion with @ahejlsberg, removed `getBestGuessSignature` in favor of just changing the behavior of `getResolvedSignature` if `!produceDiagnostics`. Also removed everything relating to `inferredAnyDefaultTypeArgument`; we will show an "inferred" type argument even if it is `{}`.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -15631,22 +15642,35 @@ namespace ts {
             //  declare function f(a: { xa: number; xb: number; });
```
comments above here need to be updated and typos fixed (specifically, that &rarr; than in one case)

sandersn commented on src/compiler/checker.ts:
```diff
@@ -15689,6 +15714,24 @@ namespace ts {
 
         }
 
+        function getBestCandidateIndex(candidates: Signature[], argsCount: number): number {
```
maybe call this `getLongestCandidateIndex`

andy-ms commented:
Unfortunately, it wouldn't be easy to break this out of the checker as we don't just need this behavior in signature help; we also need `getContextualType` to give us a good contextual signature even in the presence of errors.
